### ISSUE TITLE
[Android] Crashing when sending a message

### ISSUE THREAD

**Issue by LeagueRaINi** [mlc-ai/mlc-llm#113]
Crashes when sending a message on the exynos version of the s22 ultra on android 13, prompt length doesnt matter

<details>
  <summary>Device Info</summary>
  
```
Device
--------------------------------------------------------
Device Name : Galaxy S22 Ultra
Model : SM-S908B
Manufacturer : samsung
Device : b0s
Board : s5e9925
Hardware : s5e9925
Brand : samsung


System
--------------------------------------------------------
Android Version : 13
Version Name : T
Released Date : August 15, 2022
Code Name : Android 13
API Level : 33
One UI : 5.1
Security Patch Level : 2023-04-01
Bootloader : S908BXXU4CWCG
Baseband : S908BXXU4CWCH,S908BXXU4CWCH
Java VM : 2.1.0
Kernel : 5.10.136-android12-9-26203352-abS908BXXU4CWCG
Language : English (en_US)
OpenGL ES : 3.2
Root Management Apps : No Apps Detected
SELinux : Unable to determine
Google Play Services : 23.16.13 (190400-527363516)
Vulkan : Supported (1.1)
Treble : Supported
Seamless Updates : Not Supported


CPU
--------------------------------------------------------
SoC : Samsung Exynos 2200
     (4 nm)
1 x Cortex-X2 (2.9 GHz)
3 x Cortex-A710 (2.7 GHz)
4 x Cortex-A510 (2.1 GHz)
Processor : Samsung s5e9925
CPU Architecture : 1 x 2.80 GHz
3 x 2.52 GHz
4 x 1.82 GHz
Supported ABIs : arm64-v8a, armeabi-v7a, armeabi
CPU Hardware : s5e9925
CPU Type : 64 Bit
CPU Governor : energy_aware
Cores : 8
CPU Frequency : 400 MHz - 2803 MHz
GPU Renderer : ANGLE (Samsung Xclipse 920) on Vulkan 1.1.179
GPU Vendor : Samsung Electronics Co., Ltd.
GPU Version : OpenGL ES 3.2 ANGLE git hash: 50101db6df97


Memory
--------------------------------------------------------
Random Access Memory
------------------
Type : LPDDR5 3200 MHz
Total : 10515.0MB
Available : 4611.0MB
Used : 6065.0MB
Used Percentage : 56%

System Storage
------------------
Total : 5.33GB
Available : 0.00GB
Used : 5.33GB
Used Percentage : 99%

Internal Storage
------------------
Total : 463.02GB
Available : 371.59GB
Used : 91.43GB
Used Percentage : 19%
```
</details>

**Comment by brainer3220** [comment]
I've got this error for S21 Ultra on Android 13 too.

**Comment by brainer3220** [comment]
In addition, Seeing that it doesn't work on Pixel, I think it's because it's Exynos. 

Qualcomm device is work.

**Comment by reiuyfan** [comment]
> In addition, Seeing that it doesn't work on Pixel, I think it's because it's Exynos.
> 
> Qualcomm device is work.

It's also not working on POCO F3

**Comment by shiqimei** [comment]
Same here, my device is Huawei P30 Pro with Kirin 980 chipset, 8G RAM.

According to the logcat error, it's more like a UI level crash instead of mlc/model level.

<details>
<summary>Detailed Logs</summary>
2023-05-10 17:42:02.137 28061-28061/? I/QarthLog: [PatchStore] create disable file for ai.mlc.mlcchat uid is 11163
2023-05-10 17:42:02.148 28061-28061/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: ai.mlc.mlcchat, PID: 28061
    android.view.InflateException: Binary XML file line #27 in ai.mlc.mlcchat:layout/user_message: Binary XML file line #27 in ai.mlc.mlcchat:layout/user_message: Error inflating class <unknown>
    Caused by: android.view.InflateException: Binary XML file line #27 in ai.mlc.mlcchat:layout/user_message: Error inflating class <unknown>
    Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Constructor.newInstance0(Native Method)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:343)
        at android.view.LayoutInflater.createView(LayoutInflater.java:864)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:1016)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:971)
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:1133)
        at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:1094)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:692)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:536)
        at com.android.internal.policy.HwPhoneLayoutInflater.inflate(HwPhoneLayoutInflater.java:138)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:483)
        at ai.mlc.mlcchat.MessageAdapter.getView(MessageAdapter.java:72)
        at android.widget.AbsListView.obtainView(AbsListView.java:2662)
        at android.widget.HwAbsListView.obtainView(HwAbsListView.java:1309)
        at android.widget.ListView.measureHeightOfChildren(ListView.java:1505)
        at android.widget.ListView.onMeasure(ListView.java:1411)
        at android.view.View.measure(View.java:25166)
        at androidx.constraintlayout.widget.ConstraintLayout$Measurer.measure(ConstraintLayout.java:763)
        at androidx.constraintlayout.solver.widgets.analyzer.BasicMeasure.measure(BasicMeasure.java:426)
        at androidx.constraintlayout.solver.widgets.analyzer.BasicMeasure.measureChildren(BasicMeasure.java:105)
        at androidx.constraintlayout.solver.widgets.analyzer.BasicMeasure.solverMeasure(BasicMeasure.java:247)
        at androidx.constraintlayout.solver.widgets.ConstraintWidgetContainer.measure(ConstraintWidgetContainer.java:117)
        at androidx.constraintlayout.widget.ConstraintLayout.resolveSystem(ConstraintLayout.java:1532)
        at androidx.constraintlayout.widget.ConstraintLayout.onMeasure(ConstraintLayout.java:1607)
        at android.view.View.measure(View.java:25166)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:7109)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
        at androidx.appcompat.widget.ContentFrameLayout.onMeasure(ContentFrameLayout.java:145)
        at android.view.View.measure(View.java:25166)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:7109)
        at android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1580)
        at android.widget.LinearLayout.measureVertical(LinearLayout.java:849)
        at android.widget.LinearLayout.onMeasure(LinearLayout.java:728)
        at android.view.View.measure(View.java:25166)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:7109)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
        at android.view.View.measure(View.java:25166)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:7109)
        at android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1580)
        at android.widget.LinearLayout.measureVertical(LinearLayout.java:849)
        at android.widget.LinearLayout.onMeasure(LinearLayout.java:728)
        at android.view.View.measure(View.java:25166)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:7109)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
        at com.android.internal.policy.DecorView.onMeasure(DecorView.java:1139)
        at android.view.View.measure(View.java:25166)
        at android.view.ViewRootImpl.performMeasure(ViewRootImpl.java:3699)
        at android.view.ViewRootImpl.measureHierarchy(ViewRootImpl.java:2324)
        at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2721)
        at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:2196)
2023-05-10 17:42:02.150 28061-28061/? E/AndroidRuntime:     at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:9004)
        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1280)
        at android.view.Choreographer.doCallbacks(Choreographer.java:1019)
        at android.view.Choreographer.doFrame(Choreographer.java:911)
        at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1248)
        at android.os.Handler.handleCallback(Handler.java:900)
        at android.os.Handler.dispatchMessage(Handler.java:103)
        at android.os.Looper.loop(Looper.java:219)
        at android.app.ActivityThread.main(ActivityThread.java:8673)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:513)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1109)
     Caused by: android.content.res.Resources$NotFoundException: Drawable ai.mlc.mlcchat:drawable/user_rounded_corner with resource ID #0x7f0700dd
     Caused by: android.content.res.Resources$NotFoundException: File res/qo.xml from drawable resource ID #0x7f0700dd
        at android.content.res.ResourcesImpl.loadDrawableForCookie(ResourcesImpl.java:1314)
        at android.content.res.ResourcesImpl.loadDrawable(ResourcesImpl.java:1050)
        at android.content.res.Resources.loadDrawable(Resources.java:1004)
        at android.content.res.TypedArray.getDrawableForDensity(TypedArray.java:1195)
        at android.content.res.TypedArray.getDrawable(TypedArray.java:1170)
        at android.view.View.<init>(View.java:5431)
        at android.widget.TextView.<init>(TextView.java:1034)
        at android.widget.TextView.<init>(TextView.java:1028)
        at androidx.appcompat.widget.AppCompatTextView.<init>(AppCompatTextView.java:108)
        at androidx.appcompat.widget.AppCompatTextView.<init>(AppCompatTextView.java:103)
        at java.lang.reflect.Constructor.newInstance0(Native Method)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:343)
        at android.view.LayoutInflater.createView(LayoutInflater.java:864)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:1016)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:971)
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:1133)
        at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:1094)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:692)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:536)
        at com.android.internal.policy.HwPhoneLayoutInflater.inflate(HwPhoneLayoutInflater.java:138)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:483)
        at ai.mlc.mlcchat.MessageAdapter.getView(MessageAdapter.java:72)
        at android.widget.AbsListView.obtainView(AbsListView.java:2662)
        at android.widget.HwAbsListView.obtainView(HwAbsListView.java:1309)
        at android.widget.ListView.measureHeightOfChildren(ListView.java:1505)
        at android.widget.ListView.onMeasure(ListView.java:1411)
        at android.view.View.measure(View.java:25166)
        at androidx.constraintlayout.widget.ConstraintLayout$Measurer.measure(ConstraintLayout.java:763)
        at androidx.constraintlayout.solver.widgets.analyzer.BasicMeasure.measure(BasicMeasure.java:426)
        at androidx.constraintlayout.solver.widgets.analyzer.BasicMeasure.measureChildren(BasicMeasure.java:105)
        at androidx.constraintlayout.solver.widgets.analyzer.BasicMeasure.solverMeasure(BasicMeasure.java:247)
        at androidx.constraintlayout.solver.widgets.ConstraintWidgetContainer.measure(ConstraintWidgetContainer.java:117)
        at androidx.constraintlayout.widget.ConstraintLayout.resolveSystem(ConstraintLayout.java:1532)
        at androidx.constraintlayout.widget.ConstraintLayout.onMeasure(ConstraintLayout.java:1607)
        at android.view.View.measure(View.java:25166)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:7109)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:194)
        at androidx.appcompat.widget.ContentFrameLayout.onMeasure(ContentFrameLayout.java:145)
        at android.view.View.measure(View.java:25166)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:7109)
2023-05-10 17:42:02.167 1530-6820/? W/ActivityTaskManager:   finishTopCrashedActivityLocked Force finishing activity ai.mlc.mlcchat/.MainActivity
2023-05-10 17:42:02.178 1530-6820/? I/HwActivityTaskManagerServiceEx: call returnToHome, from: ActivityRecord{9d51687 u0 ai.mlc.mlcchat/.MainActivity t20721 f}, to: ActivityRecord{a151383 u0 com.huawei.android.launcher/.unihome.UniHomeLauncher t1}
</details>

**Comment by JackJollimore** [comment]
> Thanks for the feedbacks, we didn't get our hand on other phones, but would love to continue improve. There is a recent update to reduce the RAM usage of APK to ~4.5G (previous due to an OpenCL issue the memory usage can be unreasonably high), so if you crashed due to memory issue, feel free to try out (redownload from same apk link will work). We are aware it is not yet working on Pixel.

Hi, 

I'm not sure it is a memory issue. For context, I am able to run a various models in termux. My device is samsung s10+ qualcomm, so it's capable, but the application closes before the A.I. can respond.

I uninstalled the previous version and downloaded the newer version and experiencing the same force close/crash issue.

**Comment by 18140663659** [comment]
> 

The same issue, did you manage to solve it later?

**Comment by 18140663659** [comment]
> Same here (model is downloaded in local successfully, but crash when sending the first message), my device is Huawei P30 Pro with Kirin 980 chipset, 8G RAM.
> 
> According to the logcat error, it's more like a UI level crash instead of mlc/model level.
> 
> Detailed Logs
> TL;DR
> 
> android.view.InflateException: Binary XML file line #27 in ai.mlc.mlcchat:layout/user_message: Binary XML file line #27 in ai.mlc.mlcchat:layout/user_message: Error inflating class Caused by: android.view.InflateException: Binary XML file line #27 in ai.mlc.mlcchat:layout/user_message: Error inflating class Caused by: java.lang.reflect.InvocationTargetException at java.lang.reflect.Constructor.newInstance0(Native Method)


The same issue, did you manage to solve it later?

**Comment by bubbabug** [comment]
Samsung Galaxy Note 10 Snapdragon crashing whenever I hit send. Can't get a single message through before crash.

**Comment by jokerElsa** [comment]
same issue with samsung S10

**Comment by jokerElsa** [comment]
--------- beginning of crash
2023-05-25 14:17:16.110 451-555/? A/libc: Fatal signal 6 (SIGABRT), code -1 (SI_QUEUE) in tid 555 (pool-4-thread-1), pid 451 (ai.mlc.mlcchat)
2023-05-25 14:17:16.714 1406-1406/? A/DEBUG: *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
2023-05-25 14:17:16.714 1406-1406/? A/DEBUG: Build fingerprint: 'samsung/beyond1qltezc/beyond1q:12/SP1A.210812.016/G9730ZCU8HVL2:user/release-keys'
2023-05-25 14:17:16.714 1406-1406/? A/DEBUG: Revision: '17'
2023-05-25 14:17:16.714 1406-1406/? A/DEBUG: ABI: 'arm64'
2023-05-25 14:17:16.714 1406-1406/? A/DEBUG: Processor: '7'
2023-05-25 14:17:16.714 1406-1406/? A/DEBUG: Timestamp: 2023-05-25 14:17:16.168176132+0800
2023-05-25 14:17:16.714 1406-1406/? A/DEBUG: Process uptime: 31s
2023-05-25 14:17:16.714 1406-1406/? A/DEBUG: Cmdline: ai.mlc.mlcchat
2023-05-25 14:17:16.714 1406-1406/? A/DEBUG: pid: 451, tid: 555, name: pool-4-thread-1  >>> ai.mlc.mlcchat <<<
2023-05-25 14:17:16.714 1406-1406/? A/DEBUG: uid: 10609
2023-05-25 14:17:16.714 1406-1406/? A/DEBUG: signal 6 (SIGABRT), code -1 (SI_QUEUE), fault addr --------
2023-05-25 14:17:16.714 1406-1406/? A/DEBUG: Abort message: 'terminating with uncaught exception of type tvm::runtime::InternalError: [14:17:16] /Users/houbohan/relax/include/tvm/runtime/device_api.h:291: unknown type =0
    Stack trace not available when DMLC_LOG_STACK_TRACE is disabled at compile time.

**Comment by JFlamel** [comment]
Also not working on Samsung Note 8 with 10GB swap file.
![logcat](https://github.com/mlc-ai/mlc-llm/assets/130377245/ec983e81-b5a5-4465-ba81-d0528ccb1f59)

**Comment by fillwithjoy1** [comment]
> Thanks for the feedbacks, we didn't get our hand on other phones, but would love to continue improve. There is a recent update to reduce the RAM usage of APK to ~4.5G (previous due to an OpenCL issue the memory usage can be unreasonably high), so if you crashed due to memory issue, feel free to try out (redownload from same apk link will work). We are aware it is not yet working on Pixel.

I have a Pixel. Can I help get this working on Pixel devices?

**Comment by tqchen** [comment]
Please check out the latest apk and feel free to open new issues
